Scrape electoral history

ondenman opened the pull request:
Member electoral history is published on the individual pages of each member. (Example page: http://www.parlament.mt/joe-cassar)

The non-semantic, inconsistent way in which the data is laid out makes it difficult to scrape. This PR parses the electoral history section in to an easily manageable hash.

## Note to reviewer:

This is a step towards scraping ex members, along with their resignation dates which will address this issue (https://github.com/everypolitician/everypolitician-data/issues/24566)

## [Scraper Change checklist](https://github.com/everypolitician/everypolitician/wiki/Scraper-Change-checklist)
* [x] 1. scraper is on Morph.io under the "everypolitician-scrapers" group?
* [x] 2. scraper's GitHub "Website" link points at morph.io page?
* [x] 3. scraper is set to auto-run?
* [x] 4. scraper is archiving?
* [x] 5. legislature has a scraper webhook set?

ondenman committed:
Add rake test task

ondenman committed:
Add test helper

ondenman committed:
Test that member's electoral history has correct number of terms

ondenman committed:
Test that ex member has correct end date

ondenman committed:
Scrape electoral history to hash

ondenman committed:
Extract electoral_history logic to separate methods

tmtmtmtm commented on lib/member_page.rb:
```diff
@@ -31,26 +31,31 @@ class MemberPage < Scraped::HTML
-    terms.map { |str| str.delete(':') }
-         .map(&:tidy)
-         .zip(data).to_h
+    data = electoral_history_section.text
```
This is all much too complex to understand here. I suggest you make the electoral history section a `fragment`, so you can have different methods operating on it.

tmtmtmtm commented on tests/electoral_data_test.rb:
```diff
@@ -0,0 +1,34 @@
+  describe 'electoral history' do
+    it 'should have an end date' do
+      require 'pry'
+      subject.electoral_history['Twelfth Parliament']['Resignation from Parliament'].must_equal '3.11.15'
```
I thought one of the goals here would have been to have this equal '2015-11-03'?

tmtmtmtm reviewed (changes requested):
NB: I haven't yet checked whether this is actually working against every page. If it's not _meant_ to yet, but is only known to work against a subset for now, you should note that, and have a linked issue. If it _is_ working on all pages, then it would good to save the relevant dates. (If that's happening in a different PR, there should be an issue for _that_)

ondenman committed:
Return empty hash for electoral history

ondenman committed:
Test that current member electoral history is empty

Capturing electoral history for all members is going to be a complex task.
So, for the moment we will only implement it for ex members.

We want all members to respond to #electoral_history but we'll return an empty hash for regular members

ondenman committed:
Add ExMemberPage

Subclasses MemberPage. Returns populate :electoral_history.

ondenman committed:
Test ex member page has end date

ondenman committed:
Create ElectoralHistory fragment class

ondenman committed:
fixup! Test ex member page has end date

ondenman committed:
Add date formatter decorator

Parses response body and formats all dates into the following
format: YYY-MM-DD

ondenman committed:
Split dates by YYYY-MM-DD format

ondenman committed:
Use FormattedDates decorator

ondenman committed:
fixup! Create ElectoralHistory fragment class

tmtmtmtm commented on lib/electoral_history.rb:
```diff
@@ -0,0 +1,33 @@
+require 'scraped'
+
+class ElectoralHistory < Scraped::HTML
+  def to_h
```
clobbering the base `to_h` is almost certainly a very bad idea. Doing this essentially negates it being a `Scraped::HTML` subclass. Can you not implement this using `field`s as normal?

tmtmtmtm reviewed (changes requested):
I still haven't dug into the implementation of the electoral history here, but this is starting to turn into a monster. I'd suggest splitting this up into several smaller PRs. Adding the test framework separately (with just a simple test for a field that already exists) looks like it would help.

ondenman commented:
Addresses: https://github.com/everypolitician/everypolitician-data/issues/25279
